Replace pdoc with MkDocs

abbydiode opened the pull request:
Little bit of a messy PR, but should work. Needed some adaptations to work with MkDocs but got it working locally, might want to check workflow for tests.

MiguelGuthridge commented on .vscode/extensions.json:
```diff
@@ -1,17 +1,7 @@
 	"recommendations": [
 		"ms-python.vscode-pylance",
 		"ms-python.python",
-		"streetsidesoftware.code-spell-checker",
-		"njpwerner.autodocstring",
-		"kevinrose.vsc-python-indent"
```
Could you undo these tweaks please? I'm recommending the spell checker since it's pretty important to ensure that things are spelt correctly within the documentation. The Python indent one can probably be removed, since it's just a quality of life thing (although I'll sort that out in a separate PR).

MiguelGuthridge commented on .mypy.ini:
```diff
@@ -1 +0,0 @@
-[mypy]
```
Please keep this file. I'm planning on using it to increase type strictness in the near future.

MiguelGuthridge commented on .gitmodules:
```diff
@@ -1,3 +1,3 @@
 [submodule "src/fl_classes"]
 	path = src/fl_classes
-	url = git@github.com:MiguelGuthridge/fl_classes.git
```
Is there any particular reason why this should be changed?

abbydiode replied:
I'm not sure why that changed to be honest, I had some troubles getting the submodule to fetch on my end and I guess it changed it to the URL while I was messing around with it. Shouldn't be much of a difference but I can revert it if you'd like?

MiguelGuthridge replied:
The main reason it was SSH rather than HTTPS was since then pushing to it (which I almost always do when working on this repo) doesn't require a username and password each time. I'll investigate to see if there's some way I can get this working so that it just picks whichever protocol works properly rather than requiring one or the other.

MiguelGuthridge commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,54 @@
+          
+      - name: Deploy
+        id: deploy
+        uses: actions/deploy-pages@v1
```
More of a note to self, but this will deploy to a different URL, breaking existing links. I need to do some redirects and stuff to make this all work.

abbydiode replied:
To clarify, this will deploy without relying on a branch. In the repository's settings for pages it should be set to GitHub Actions if you weren't already using that. The URL should still be https://miguelguthridge.github.io/FL-Studio-API-Docs/ though.

MiguelGuthridge replied:
Neat, if that's correct, this should be a huge improvement on the massively convoluted system I was using before (literally using API keys to copy stuff from one repo to another that gets statically hosted). I can't see any definition in the file though. Are you sure it won't just deploy to whatever the repo name is?

abbydiode replied:
I'll add comments to the workflow file, to make it easier for both of us to understand what's going on.

MiguelGuthridge commented on .vscode/settings.json:
```diff
@@ -1,28 +1,5 @@
 {
-    // "autoDocstring.docstringFormat": "numpy",
-    "autoDocstring.customTemplatePath": "resources/docstring_template.mustache",
-    "cSpell.language": "en",
-    "cSpell.words": [
-        "automatable",
-        "autouse",
-        "BBGGRR",
-        "deinitialize",
-        "desynchronize",
-        "docformat",
-        "Enveditor",
-        "falsiness",
-        "indexedness",
-        "jsonschema",
-        "MIDICC",
-        "mypy",
-        "pdoc",
-        "ppqn",
-        "Precount",
-        "pyscript",
-        "Rect",
-        "Sysex",
-        "timebase"
-    ],
```
Please undo these modifications, since I intend to keep the spell checker enabled.

MiguelGuthridge commented on docs/scripts/generate_pages.py:
```diff
@@ -0,0 +1,16 @@
+src = Path("src")
+modules = []
+
+for path in src.rglob("*.py"):
+    module = Path(str(path.relative_to(src)).replace("__", "")).parent
+    if module not in modules:
+        modules.append(module)
+
+for module in modules:
+    module_path = (Path(module, "index") if any(m.parent == module for m in modules) else module).with_suffix(".md")
+    with mkdocs_gen_files.open(module_path, "w") as f:
+        identifier = ".".join(module.parts)
+        print(f"::: {identifier}", file=f)
```
I'd appreciate a few comments explaining what's going on here, since that should hopefully make it more maintainable in the future.

MiguelGuthridge commented on resources/Release Checklist.md:
```diff
@@ -1,12 +0,0 @@
-
-# Release Checklist
-
-* [ ] Check that CI passed
-
-* [ ] Update version number
-
-* [ ] Update `fl_model/configuration/schema.json` for major versions
-
-* [ ] Update `LATEST_API_VERSION` from `fl_model/consts.py` for major versions
-
-* [ ] Write release notes and release
```
I'd prefer to keep this here for the time being, since I want to ensure that I don't miss any important steps when releasing new versions.

MiguelGuthridge reviewed (changes requested):
This is looking like it's a great start, but there are a few things that have changed that I'd prefer don't change, and a few others that are probably best saved for other PRs. I'd appreciate it if you can undo those changes, so that we can get these awesome improvements merged!

abbydiode commented:
Sorry, couple of accidental removals in there before. As I mentioned I had no intention to merge back to the main repository so I completely changed the repository at first, your requested changes should be handled now.

MiguelGuthridge reviewed (commented):
Looking pretty good! I'll do a bit more testing of this over the next week or so (am busy traveling from tomorrow onwards and need to pack tonight), but this honestly looks pretty awesome! I haven't run the scripts on my machine yet, so I might find a few bugs, but considering it was working for you this is unlikely. I'm thinking you probably won't need to do anything more to get this merged, but I'll let you know if I run into any issues. Thanks again for all your work! This is genuinely such a massive improvement over the old online docs!

abbydiode commented:
> Looking pretty good! I'll do a bit more testing of this over the next week or so (am busy traveling from tomorrow onwards and need to pack tonight), but this honestly looks pretty awesome! I haven't run the scripts on my machine yet, so I might find a few bugs, but considering it was working for you this is unlikely. I'm thinking you probably won't need to do anything more to get this merged, but I'll let you know if I run into any issues. Thanks again for all your work! This is genuinely such a massive improvement over the old online docs!

I've added the requested comments and reverted the deletions you wanted me to revert. Funnily enough I'm traveling too and leaving today so this will most likely be the last from me for a few weeks now. Enjoy your holiday!

MiguelGuthridge commented:
One other thing I'm a little unsure about is how the main page is pulled from `docs/index.md`. I couldn't spot anything that seemed to explicitly include it. 

I think that if it's easy to link in plain old markdown like this, it would be a great idea to write some comprehensive documentation on aspects of the API that can't really be attributed to any particular function or module, for example how to set up a new script, or how to use things like REC events (this info is currently in the device module docs which I feel is a little awkward).

abbydiode commented:
> One other thing I'm a little unsure about is how the main page is pulled from `docs/index.md`. I couldn't spot anything that seemed to explicitly include it.

MkDocs magic. It just treats that file as the one for the main page when it builds. 

> I think that if it's easy to link in plain old markdown like this, it would be a great idea to write some comprehensive documentation on aspects of the API that can't really be attributed to any particular function or module, for example how to set up a new script, or how to use things like REC events (this info is currently in the device module docs which I feel is a little awkward).

Setting up new static pages is really quite simple and very useful for something like a guide. Check out this example page I just made in less than a minute.

https://github.com/abbydiode/fl-studio-api-docs/blob/main/docs/example.md -> https://abbydiode.github.io/fl-studio-api-docs/example/

MiguelGuthridge commented:
Ok I've tried it out, and it looks pretty awesome! One minor concern I have is with the way that the docs are built into the already-existing `docs/` directory, where they aren't git-ignored. Would it be possible to have the existing docs stuff be copied to a directory such as `docs_build` with the `generate_pages` script then populating the remaining files to that directory. That way, that folder can be gitignored which will simplify the process of building the docs locally, without risking messing up Git or inadvertently overwriting work inside the `docs` folder (eg what happens if Image-Line added an `index` module - that could overwrite our existing `index.md` file).

abbydiode commented:
> Ok I've tried it out, and it looks pretty awesome! One minor concern I have is with the way that the docs are built into the already-existing `docs/` directory, where they aren't git-ignored. Would it be possible to have the existing docs stuff be copied to a directory such as `docs_build` with the `generate_pages` script then populating the remaining files to that directory. That way, that folder can be gitignored which will simplify the process of building the docs locally, without risking messing up Git or inadvertently overwriting work inside the `docs` folder (eg what happens if Image-Line added an `index` module - that could overwrite our existing `index.md` file).

When testing locally it's best to use `mkdocs serve`, this will serve the site on a simple local webserver. When the documentation is actually built using `mkdocs build` it builds into the `./site/` directory, which is gitignored. The GitHub pages deployment workflow takes this directory and creates an artifact that can be used by GitHub pages directly.

Unless I've misunderstood you, I don't think what you're mentioning is a concern. The `./docs/` directory is a source folder and is only used for MkDocs to build the site, and is static after building and does not serve from that directory anymore. The page generation just generates pages internally and serves them to MkDocs without actually creating any files.

Sorry for any confusion. Perhaps I could add some more documentation about MkDocs to the PR?

MiguelGuthridge commented:
Sorry for the confusion, to clarify, I'm referring to the `docs/scripts/generate_pages` script, which adds a bunch of files to the `docs` folder. I've checked, and it overwrites existing files in that directory, which could result in data loss. Here is my proposal for how I'd implement this.

1. Hand-written markdown documentation written in `docs` directory
2. Upon running the `generate_pages` script, the `docs` directory is copied to a new directory `docs_build` (which is gitignored)
3. Script generates the code documentation templates into the `docs_build` directory. If there are conflicting files, the script exits with an error
4. This can then be built into the `site` directory

This would ensure that the handwritten documentation isn't intermingled with script-generated documentation, meaning that writing extra documentation is simpler.

I'm gonna try implementing this myself now.

MiguelGuthridge commented:
I've given this a go. It looks like Mkdocs doesn't have support for building from multiple directories, so I've made the script merge the `docs` directory properly. This has the small limitation that it needs to merge them before `mkdocs` is run, otherwise an error is given due to the folder not existing. I modified the workflow to do this automatically. Can you think of any way to do this more cleanly?

abbydiode commented:
I'm currently unable to look into it I'm afraid. I was convinced `mkdocs-gen-pages` created them internally without overwriting any files and hadn't found any issues with it yet. I'll look into it when I get the chance.

MiguelGuthridge commented:
I've done a bit more testing and it all seems to work well enough - if there are any issues we should be able to fix them up in future PRs. Thanks again for all your work!!!
